ENH: Add support to plot 🐼s DataFrame objects

ElDeveloper opened the pull request:
This PR adds a helper function, `scatterplot`, that can take as only argument a
Pandas DataFrame and returns an Emperor object. The function itself
extracts the numerical data from the DataFrame and creates a dummy
OrdinationResults object.

Note that this is a rather experimental feature, so testing/suggestions are
very welcome.

This PR depends on #516 so merge that PR first.

mortonjt commented on emperor/pandas.py:
```diff
@@ -0,0 +1,116 @@
+        Pandas DataFrame with the data to display
+    x, y, z : str
+        Column names in `df`, these will represent the order of the axes in the
+        final visualization.
```
Can we add a comment here about why these parameters for `x`, `y`, `z` are optional?  I'm assuming that they will take the first 3 columns right?

mortonjt commented on emperor/pandas.py:
```diff
@@ -0,0 +1,116 @@
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Pandas DataFrame with the data to display
```
Alright, I'm a little confused here.

Does this dataframe contain everything?  Metadata and PC coordinates?

I wouldn't think this would be desirable - our OTU tables and metadata are often decoupled, and this would put quite a 
bit of burden on the user to get this sorted out.

In the [canvas biplot code](https://github.com/mortonjt/canvas/blob/master/canvas/biplot.py) we keep all of the metadata and PC coordinates separate.  And I think this greatly enhances useability.

ElDeveloper replied:
@mortonjt, I think I'm guilty for the confusion. Let me try to expand,
and then let's figure out how to make this clearer in the documentation.
What I was trying to do with this function, `scatterplot`, is to provide
a interface for someone to plot environmental data alone i.e. just a
mapping file. Hopefully opening up the user-base to other fields that
don't necessarily work with ordinations. Does this make sense?

On (Jul-10-16|22:58), Jamie Morton wrote:
> > +from __future__ import division
> > +
> > +import numpy as np
> > +import pandas as pd
> > +
> > +from emperor.core import Emperor
> > +from skbio import OrdinationResults
> > +
> > +
> > +def scatterplot(df, x=None, y=None, z=None, remote=True):
> > +    """Create an Emperor scatter plot from a Pandas DataFrame
> > +
> > +    Parameters
> > +    ----------
> > +    df : pd.DataFrame
> > +        Pandas DataFrame with the data to display
> 
> Alright, I'm a little confused here.
> 
> Does this dataframe contain everything?  Metadata and PC coordinates?
> 
> I wouldn't think this would be desirable - our OTU tables and metadata are often decoupled, and this would put quite a 
> bit of burden on the user to get this sorted out.
> 
> In the [canvas biplot code](https://github.com/mortonjt/canvas/blob/master/canvas/biplot.py) we keep all of the metadata and PC coordinates separate.  And I think this greatly enhances useability.
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517/files/911c1dc15f3025f6d79dc65dee8bec576c7599d9#r70205258

mortonjt commented on emperor/pandas.py:
```diff
@@ -0,0 +1,116 @@
+    # match up the metadata and coordinates
+    df = df.loc[samples.index].copy()
+
+    ores = OrdinationResults(short_method_name='Ax', long_method_name='Axis',
```
What do you think about passing in `short_method_name` and `long_method_name` as optional parameters?

ElDeveloper replied:
Sounds like a good idea, though I'm really not sure if we should even be
exposing/passing these arguments, mainly because this is not intended to
be used as an "ordination" interface ... but now, I don't know, perhaps
the best would be to just not pass anything here(?).

On (Jul-10-16|23:00), Jamie Morton wrote:
> > +    # sort columns by variance
> > +    variance = samples.var().sort_values(ascending=False)
> > +    samples = samples[variance.index].copy()
> > +
> > +    # re-order x, y and z
> > +    ordered = samples.columns.tolist()
> > +    for col in [z, y, x]:
> > +        if col is not None:
> > +            ordered.remove(col)
> > +            ordered = [col] + ordered
> > +    samples = samples[ordered].copy()
> > +
> > +    # match up the metadata and coordinates
> > +    df = df.loc[samples.index].copy()
> > +
> > +    ores = OrdinationResults(short_method_name='Ax', long_method_name='Axis',
> 
> What do you think about passing in `short_method_name` and `long_method_name` as optional parameters?
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517/files/911c1dc15f3025f6d79dc65dee8bec576c7599d9#r70205320

mortonjt commented on emperor/pandas.py:
```diff
@@ -0,0 +1,116 @@
+    # HACK: scale the position of the samples to fit better within the screen
+    ores.samples = ores.samples / ores.samples.max(axis=0)
+
+    return Emperor(ores, df, dimensions=8, remote=remote)
```
Any reason why there are `dimensions=8`?

ElDeveloper replied:
Not a good reason, I chose that arbitrarily, let me fix this.

On (Jul-10-16|23:00), Jamie Morton wrote:
> > +            ordered = [col] + ordered
> > +    samples = samples[ordered].copy()
> > +
> > +    # match up the metadata and coordinates
> > +    df = df.loc[samples.index].copy()
> > +
> > +    ores = OrdinationResults(short_method_name='Ax', long_method_name='Axis',
> > +                             eigvals=np.zeros_like(samples.columns),
> > +                             samples=samples, proportion_explained=variance)
> > +
> > +    df.index.name = '#SampleID'
> > +
> > +    # HACK: scale the position of the samples to fit better within the screen
> > +    ores.samples = ores.samples / ores.samples.max(axis=0)
> > +
> > +    return Emperor(ores, df, dimensions=8, remote=remote)
> 
> Any reason why there are `dimensions=8`?  
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517/files/911c1dc15f3025f6d79dc65dee8bec576c7599d9#r70205349

mortonjt commented:
Thanks for putting this together!  This is a very good start.

ElDeveloper commented:
Thanks so much for reviewing. I've pushed a few changes, let me know
what you think. I'm eager to be easier to read and understand 👓

On (Jul-10-16|23:01), Jamie Morton wrote:
> Thanks for putting this together!  This is a very good start.
> 
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517#issuecomment-231648749

antgonza commented:
Can you update and fix error?

mortonjt commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,102 @@
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Pandas DataFrame with the data to display
```
Don't forget to note that this also includes metadata as well as the PC coordinates.

ElDeveloper replied:
Great catch!

On (Jul-11-16|12:08), Jamie Morton wrote:
> > +from __future__ import division
> > +
> > +import numpy as np
> > +import pandas as pd
> > +
> > +from emperor.core import Emperor
> > +from skbio import OrdinationResults
> > +
> > +
> > +def scatterplot(df, x=None, y=None, z=None, remote=True):
> > +    """Create an Emperor scatter plot from a Pandas DataFrame
> > +
> > +    Parameters
> > +    ----------
> > +    df : pd.DataFrame
> > +        Pandas DataFrame with the data to display
> 
> Don't forget to note that this also includes metadata as well as the PC coordinates.
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517/files/47d735bcbfbd8fc467948a139261bdf5a7092d82#r70317772

mortonjt commented:
One comment still needs to be addressed.  But I think this will be ok otherwise.

ElDeveloper commented:
I believe I've added all that's needed. I caught a few other kinks when
I built the documentation i.e. `make -C doc html`, but I think I've
addressed them all.

On (Jul-11-16|12:10), Jamie Morton wrote:
> One comment still needs to be addressed.  But I think this will be ok otherwise.
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517#issuecomment-231833966

mortonjt commented:
Cool  👍  on my end.

antgonza commented on doc/source/index.rst:
```diff
@@ -43,3 +24,15 @@ Format Description:
+Citing Emperor
+==============
+
+Remember to cite Emperor If you use it in any published research, please
```
If -> if

antgonza commented on emperor/__init__.py:
```diff
@@ -1,3 +1,29 @@
+Emperor 3D PCoA viewer (:mod:`emperor`)
+============================================
+
+This module provides an objects and functions to interact and visualize a set
```
an objects -> objects?

antgonza commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,103 @@
+    ----------
+    df : pd.DataFrame
+        Pandas DataFrame with the data to display, this includes both
+        *metadata* and coordinates to position the samples in a 3D space.
```
why *metadata* and not *coordinates*?

mortonjt replied:
It contains both metadata and coordinates within the same dataframe.  Right
@ElDeveloper?

On Mon, Jul 11, 2016 at 12:21 PM, Antonio Gonzalez <notifications@github.com
> wrote:

> In emperor/_pandas.py
> <https://github.com/biocore/emperor/pull/517#discussion_r70319694>:
>
> > +
> > +import numpy as np
> > +import pandas as pd
> > +
> > +from emperor.core import Emperor
> > +from skbio import OrdinationResults
> > +
> > +
> > +def scatterplot(df, x=None, y=None, z=None, remote=True):
> > +    """Create an Emperor scatter plot from a Pandas DataFrame
> > +
> > +    Parameters
> > +    ----------
> > +    df : pd.DataFrame
> > +        Pandas DataFrame with the data to display, this includes both
> > +        *metadata* and coordinates to position the samples in a 3D space.
>
> why *metadata* and not *coordinates*?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/emperor/pull/517/files/508540e0df1c5599072fc3a73848b2e9ff3e7795#r70319694>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AD_a3XtOnAXkG5C1FboNxpgY0P5_e8zeks5qUpe_gaJpZM4JJDNq>
> .
>

ElDeveloper replied:
Fixed @antgonza. @mortonjt, you are correct, though I think @antgonza referred to the formatting of the words, I italicized metadata but I didn't coordinates. 🇮🇹

antgonza commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,103 @@
+
+    Notes
+    -----
+    If a row has missing data, that data pont will be removed from the
```
pont -> point

antgonza commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,103 @@
+    if not isinstance(df, pd.DataFrame):
+        raise ValueError("The argument is not a Pandas DataFrame")
+
+    for col in [z, y, x]:
```
Should you validate that x, y, z actually exist in the df?

antgonza replied:
Just saw the validation below.

antgonza commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,103 @@
+        raise ValueError("The argument is not a Pandas DataFrame")
+
+    for col in [z, y, x]:
+        if col is None:
```
How can col be None? Should you be checking for [None, None, None] or any()?

ElDeveloper replied:
These can all be `None`, if they are not explicitly specified, then the columns with the highest variance will take precedence and take the place of `x`, `y` or `z` correspondingly. That's why I chose to make them optional arguments.

antgonza replied:
Got it, thanks. Was reading that line wrongly.

antgonza commented on emperor/_pandas.py:
```diff
@@ -0,0 +1,103 @@
+    samples = samples[ordered].copy()
+
+    # match up the metadata and coordinates
+    df = df.loc[samples.index].copy()
```
Out of curiosity, why do you need copy? Basically, what's the difference between `df = df.loc[samples.index].copy()` and `df = df.loc[samples.index]`?

ElDeveloper replied:
One represents a "view" of the original data (without using copy) and the other one is a hard copy of the DataFrame (using copy). IIRC, views cannot, or should not, be modified. When you create a copy, you are explicitly creating a new object that's independent of the object from which the frame originated.

antgonza replied:
That kind of makes sense but when you do: `df = df.loc[samples.index].copy()` it should be the same effect than `df = df.loc[samples.index]`, right? I agree that this will make more sense if it was `df2 = df.loc[samples.index].copy()`.

http://stackoverflow.com/a/27680109

ElDeveloper replied:
You are right! There's no need to make a copy since those two objects
are being overwritten, I've fixed that where needed :+1:.

On (Jul-11-16|12:43), Antonio Gonzalez wrote:
> > +        raise ValueError("Not enough data to plot")
> > +
> > +    # sort columns by variance
> > +    variance = samples.var().sort_values(ascending=False)
> > +    samples = samples[variance.index].copy()
> > +
> > +    # re-order x, y and z
> > +    ordered = samples.columns.tolist()
> > +    for col in [z, y, x]:
> > +        if col is not None:
> > +            ordered.remove(col)
> > +            ordered = [col] + ordered
> > +    samples = samples[ordered].copy()
> > +
> > +    # match up the metadata and coordinates
> > +    df = df.loc[samples.index].copy()
> 
> That kind of makes sense but when you do: `df = df.loc[samples.index].copy()` it should be the same effect than `df = df.loc[samples.index]`, right? I agree that this will make more sense if it was `df2 = df.loc[samples.index].copy()`.
> 
> http://stackoverflow.com/a/27680109
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/517/files/508540e0df1c5599072fc3a73848b2e9ff3e7795#r70323071

antgonza commented:
A few comments/questions.

ElDeveloper commented:
Comments should be addressed, let's just wait for the tests to pass.
